src/engine/Server.cpp

```diff
@@ -789,31 +790,31 @@ boost::asio::awaitable<void> Server::processQuery(
 }
 
 // _____________________________________________________________________________
-template <typename Function, typename T>
+template <std::invocable Function, typename T>
 Awaitable<T> Server::computeInNewThread(Function function,
                                         SharedCancellationHandle handle) {
-  auto runOnExecutor =
-      [](auto executor, Function func,
-         SharedCancellationHandle handle) -> net::awaitable<T> {
-    co_await net::post(net::bind_executor(executor, net::use_awaitable));
-    // It might take some time until the thread pool is ready,
-    // so reset the state here. Ideally waiting for the thread pool
-    // would periodically check the cancellation state
+  auto inner = [function = std::move(function),
+                handle = std::move(handle)]() mutable -> T {
     handle->resetWatchDogState();
-    co_return std::invoke(func);
+    return std::invoke(std::move(function));
   };
-  return ad_utility::resumeOnOriginalExecutor(runOnExecutor(
-      threadPool_.get_executor(), std::move(function), std::move(handle)));
+  return ad_utility::runFunctionOnExecutor(
+      threadPool_.get_executor(), std::move(inner), net::use_awaitable);
 }
 
 // _____________________________________________________________________________
-net::awaitable<Server::PlannedQuery> Server::parseAndPlan(
+net::awaitable<std::optional<Server::PlannedQuery>> Server::parseAndPlan(
     const std::string& query, QueryExecutionContext& qec,
     SharedCancellationHandle handle, TimeLimit timeLimit) {
   auto handleCopy = handle;
+
+  // The usage of an `optional` here is required because of a limitation in
+  // Boost::Asio which forces us to use default-constructible result types with
+  // `computeInNewThread`.
   return computeInNewThread(
```

Comment:
Thoughts about unwrapping the optional inside this function? I dislike that the function signature suggests an optional that is not really optional.